### PR TITLE
New definition syntax with leading DEF keyword

### PR DESCRIPTION
**Pull request by Rangi42**
This will enable fixing #457 later once the old definition syntax is removed.

### CODE REVIEW

**Comment by Rangi42** [comment]
The `def_id` and `redef_id` rules involve an extra `strcpy`, but they're necessary to avoid a reduce/reduce conflict, since otherwise multiple rules (`def_equ`, `def_equs`, `def_set`, etc) would start with the same sequence of tokens + mid-rule actions, and bison can't tell which mid-rule to take.

**Comment by Rangi42** [comment]
How about an 0.5.0-rc2 build in the next few days or so after this is merged? (That way I can add `DEF` to all my constants for more testing.)

File: src/asm/parser.y
Comment (ISSOtm) [inline src/asm/parser.y:699]: I think that this could be "either `DEF <ID>` or `<LABEL>`", to avoid duplicating all symbol definition productions.

It *might* cause parser conflicts, though? And maybe it doesn't make sense, and we'd better stick to this? I'll let you judge, I don't have a strong opinion.

File: src/asm/rgbasm.5
Comment (ISSOtm) [inline src/asm/rgbasm.5:1129]: Unlike numeric symbols, this doesn't mention that EQUS are expanded using the old syntax.

File: src/asm/rgbasm.5
Comment (ISSOtm) [inline src/asm/rgbasm.5:1636]: In general, I think we should say that they are not *implicitly* expanded, since `{SYM}` does work. Maybe the documentation for `EQUS` expansion should also mention the word "explicitly", to prime this explanation.

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/asm/parser.y
Comment (Rangi42) [reply src/asm/parser.y:699]: The new `def_id EQU/EQUS/SET/etc` directives fall under the `| label directive` pattern of `plain_directive`s, whereas old-style `label EQU/EQUS/SET/etc` are `assignment_directive`s. Also, in future the old-style ones will act differently, defining the constant but also printing a deprecation warning.

Comment (ISSOtm) [reply src/asm/parser.y:699]: We'll see whether we deprecate the old syntax; it should still be useful for compatibility, doesn't cost anything, and the new syntax is there to avoid the newbie trap.

File: src/asm/rgbasm.5
Comment (Rangi42) [reply src/asm/rgbasm.5:1636]: I had taken "string symbols are expanded" to always be talking about `EQUS` expansions. `{This}` is consistently called "interpolation", even outside string literals.

Comment (Rangi42) [reply src/asm/rgbasm.5:1637]: (This also changes a mention of "`EQUS` symbols" to "string symbols", consistent with other mentions, and to go with the definition: "`EQUS` is used to define string symbols." So just searching for "string symbol" if it's not clear will turn up what this is talking about.)

Comment (ISSOtm) [reply src/asm/rgbasm.5:1637]: I think it's better to have a little redundancy to reinforce the difference between the two, to further help prevent a reader skimming the document from conflating the two. I don't think that it hurts readability, but I'm no writer.

Comment (Rangi42) [reply src/asm/rgbasm.5:1637]: To avoid repeating the same info everywhere, how about saying in the explanation of `{interpolation}` that it will occur "even in contexts where string symbols are not expanded, like `PURGE name`, `DEF(name)`, or `MACRO name`"?

Comment (Rangi42) [reply src/asm/rgbasm.5:1637]: Also I can make the term consistently "string equate" instead.

Comment (ISSOtm) [reply src/asm/rgbasm.5:1637]: Yes please to both!

Comment (Rangi42) [reply src/asm/rgbasm.5]: Done :)

File: src/asm/rgbasm.5
Comment (ISSOtm) [inline src/asm/rgbasm.5:1023]: I think we should explicit that it's as soon as the EQUS' *name* is encountered, that the expansion takes place.

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/asm/rgbasm.5
Comment (Rangi42) [reply src/asm/rgbasm.5:1023]: How about "the symbol's name is expanded as if its contents had been there instead." ?

Comment (ISSOtm) [reply src/asm/rgbasm.5:1023]: Perhaps "replaced" instead of "expanded"? Otherwise, yes.

Comment (Rangi42) [reply src/asm/rgbasm.5]: Okay.

**Review by ISSOtm — APPROVED** [review]

**Pull request merged**
